sw/device/tests/sim_dv/spi_tx_rx_test.c

```diff
@@ -0,0 +1,385 @@
+         fired_irqs[kDifSpiDeviceIrqTypeRxFull];
+}
+
+static bool execute_test(const dif_spi_device_t *spi_device) {
+  LOG_INFO("Executing the test.");
+
+  size_t bytes_transferred = 0;
+  CHECK(dif_spi_device_send(spi_device, spi_device_tx_data,
+                            SPI_DEVICE_DATASET_SIZE,
+                            &bytes_transferred) == kDifSpiDeviceResultOk);
+  if (bytes_transferred != SPI_DEVICE_DATASET_SIZE) {
+    LOG_ERROR(
+        "SPI_DEVICE TX_FIFO transferred bytes mismatched: {act: %0d, exp: %0d}",
+        bytes_transferred, SPI_DEVICE_DATASET_SIZE);
+  } else {
+    LOG_INFO("Transferred %0d bytes to SPI_DEVICE TX_FIFO.", bytes_transferred);
+  }
+
+  CHECK(dif_spi_device_set_irq_levels(spi_device, SPI_DEVICE_DATASET_SIZE,
+                                      SPI_DEVICE_DATASET_SIZE / 2) ==
+        kDifSpiDeviceResultOk);
+  expected_irqs[kDifSpiDeviceIrqTypeTxBelowLevel] = true;
+
+  bool read_rx_fifo_done = false;
+  while (!read_rx_fifo_done || !exp_irqs_fired()) {
+    // set rx tx level back to default value so TxBelowLevel irq won't trigger
+    if (fired_irqs[kDifSpiDeviceIrqTypeTxBelowLevel] &&
+        expected_irqs[kDifSpiDeviceIrqTypeTxBelowLevel]) {
+      CHECK(dif_spi_device_set_irq_levels(spi_device, SPI_DEVICE_DATASET_SIZE,
+                                          0) == kDifSpiDeviceResultOk);
+      expected_irqs[kDifSpiDeviceIrqTypeTxBelowLevel] = false;
+      expected_irqs[kDifSpiDeviceIrqTypeRxAboveLevel] = true;
+      LOG_INFO("SPI_DEVICE tx_below_level interrupt fired.");
+    }
+
+    // wait for SPI_HOST to send 128 bytes and trigger RxAboveLevel irq
+    if (fired_irqs[kDifSpiDeviceIrqTypeRxAboveLevel] &&
+        expected_irqs[kDifSpiDeviceIrqTypeRxAboveLevel]) {
+      expected_irqs[kDifSpiDeviceIrqTypeRxAboveLevel] = false;
+      LOG_INFO("SPI_DEVICE rx_above_level interrupt fired.");
+    }
+
+    // when 128 bytes received in RX_FIFO from SPI_HOST,
+    // read out and compare against the expected data
+    if (fired_irqs[kDifSpiDeviceIrqTypeRxAboveLevel] && !read_rx_fifo_done) {
+      size_t bytes_recved = 0;
+      uint8_t spi_device_rx_data[SPI_DEVICE_DATASET_SIZE];
+      CHECK(dif_spi_device_recv(spi_device, spi_device_rx_data,
+                                SPI_DEVICE_DATASET_SIZE,
+                                &bytes_recved) == kDifSpiDeviceResultOk);
+      if (bytes_recved == SPI_DEVICE_DATASET_SIZE) {
+        LOG_INFO("Received %0d bytes from SPI_DEVICE RX_FIFO.", bytes_recved);
+        read_rx_fifo_done = true;
+      } else {
+        LOG_ERROR(
+            "SPI_DEVICE RX_FIFO recvd bytes mismatched: {act: %0d, exp: %0d}",
+            bytes_recved, SPI_DEVICE_DATASET_SIZE);
+      }
+      // expect SPI_HOST to send another 1024 bytes to fill RX SRAM FIFO
+      expected_irqs[kDifSpiDeviceIrqTypeRxAboveLevel] = true;
+      expected_irqs[kDifSpiDeviceIrqTypeRxFull] = true;
+      expected_irqs[kDifSpiDeviceIrqTypeTxUnderflow] = true;
```

Comment:
is it safer to expect these interrupts one by one in case there are some misconnections.

kDifSpiDeviceIrqTypeTxUnderflow should occur first and repeat for every transaction, and then kDifSpiDeviceIrqTypeRxAboveLevel, which occurs once, kDifSpiDeviceIrqTypeRxFull will be the last

Reply:
that is a good point, thanks for the suggestion. Let me try to separate them.